### ISSUE TITLE
OAuth & Getting Started instructions do not work

### ISSUE THREAD

**Issue by Capstan** [google/google-api-go-client#133]
TL;DR: The version of `golang.org/x/oauth2` you get from `go get` does not work with any of the stated documentation.

~~1. [README.md](https://github.com/google/google-api-go-client/blob/master/README.md) refers to issuing `go get google.golang.org/api/storage/v1` and gives an example of Application Default Credentials. These examples and their docs simply do not work because the version of golang/x/oauth2 etc. that you get with `go get` do not contain support for Application Default Credentials' methods.~~
1. [GettingStarted.md's OAuth HTTP Client instructions](https://github.com/google/google-api-go-client/blob/master/GettingStarted.md#oauth-http-client) points to [example code](https://github.com/google/google-api-go-client/tree/master/examples) that don't do any credential handling, but instead refer to a unnamed `registerDemo` that is not findable using GitHub search. Grepping finally finds it in `examples/main.go`. Perhaps the link should go to main.go instead?
1. [GettingStarted.md's OAuth HTTP Client instructions](https://github.com/google/google-api-go-client/blob/master/GettingStarted.md#oauth-http-client) points to an OAuth config that won't work, even if you do figure out you need to call `oauth2.Config.AuthCodeUrl` because no `RedirectURL` was provided, and unless you're going to erect a whole new server like `examples/main.go` does, you need to [choose a redirect URI](https://developers.google.com/identity/protocols/OAuth2InstalledApp#choosingredirecturi), and the simplest form for a standalone command-line binary is `urn:ietf:wg:oauth:2.0:oob`. Finding how to do this was not straightforward.
1. [GettingStarted.md's OAuth HTTP Client instructions](https://github.com/google/google-api-go-client/blob/master/GettingStarted.md#oauth-http-client) refer to a `newOAuthClient` function that is not of a package, but then doesn't tell you what package it's in and has no link. (And actually, it's not in any package, but is a part of `examples/main.go`, which you cannot include.)

~~Finally, GoDoc for [golang.org/x/oauth2](https://godoc.org/golang.org/x/oauth2),  [golang.org/x/oauth2/google](https://godoc.org/golang.org/x/oauth2/google), and even [google.golang.api/storage/v1](https://godoc.org/google.golang.org/api/storage/v1) all refer to newer code than what `go get` fetches, so it leads you down the primrose path. Is there a GoDoc docset frozen to what `go get` fetches? (e.g., with a version identifier.)~~

**Comment by bradfitz** [comment]
Did you use `go get -u`? I suspect you ran only `go get` without `-u` and it just used the old version you already had on disk.

**Comment by Capstan** [comment]
No. I did not. I looked for some time to figure out if `go get` had some update flag, but clearly overlooked it. I could not find anything that said what the version of what I had checked out was or whether it was out of date. :/ Alright, now that I've done that, I'll try this all again.

**Comment by Capstan** [comment]
Would it be unreasonable to add "-u" to the getting started flags? I obviously had a stale state and no amount of repeating the instructions would get me to working.

**Comment by Capstan** [comment]
OK, so Application Default creds don't work for me, since I'm trying to use user creds, so I'm back to the original issues.

**Comment by mcgreevy** [comment]
@Capstan can you paste the (non-working) code that you have so far?

**Comment by Capstan** [comment]
This code was cribbed from a StackOverflow post before I ever found `main.go`:

```go
package main

import (
  "fmt"
  "log"

  "golang.org/x/net/context"
  "golang.org/x/oauth2"
  "golang.org/x/oauth2/google"
  "google.golang.org/api/storage/v1"
)

func main() {
  // boilerplate
  var config = &oauth2.Config{
    ClientID:     "<redacted>", // from https://console.developers.google.com/project/<your-project-id>/apiui/credential
    ClientSecret: "<redacted>", // from https://console.developers.google.com/project/<your-project-id>/apiui/credential
    Endpoint:     google.Endpoint,
    RedirectURL:  "urn:ietf:wg:oauth:2.0:oob",
    Scopes:       []string{storage.DevstorageFullControlScope},
  }
  
  url := config.AuthCodeURL("state", oauth2.AccessTypeOnline)
  fmt.Printf("Visit the URL for the auth dialog: %v\nEnter it here: ", url)

  var code string
  if _, err := fmt.Scan(&code); err != nil {
    log.Fatal(err)
  }
  tok, err := config.Exchange(oauth2.NoContext, code)
  if err != nil {
    log.Fatal(err)
  }
  httpClient := config.Client(context.Background(), tok)
  if err != nil {
    fmt.Printf("Could not create http client: %v", err)
  }

  svc, err := storage.New(httpClient)
  if err != nil {
    fmt.Printf("Could not create storage client: %v", err)
  }
  
  var _ = svc
  fmt.Printf("Constructed a service!\n")
}
```

This actually functions, but it 1) isn't anything that is pointed to by the help, and 2) doesn't do any of the bells and whistles of token caching and erecting a local http service to allow redirection.

Is there no equivalent of Application Default creds' simplicity in a public packaged `newOAuthClient` that does all the creds saving for you? Or is picking the local client id file and token cache file all left as "have the user figure it out"?

**Comment by bradfitz** [comment]
Sent a CL to update the docs a bit: https://code-review.googlesource.com/#/c/4241/

**Comment by mcgreevy** [comment]
FYI I am going to reassign this to broady, but need to sort out some project settings problems first.

**Comment by eladg** [comment]
while we are at it, how about we also include an example for non OAuth api call?

```go
client := &http.Client{
  Transport: &transport.APIKey{Key: GOOGLE_API_SERVER_KEY},
}
service, err := google_service.New(client)
```

It's fairly basic, but might really help beginners.

**Comment by broady** [comment]
@eladg good idea - not many APIs use this auth mechanism, but would be good to have that in there.

**Comment by kamronbatman** [comment]
I have been trying to get my appengine instance (not compute engine) to work with Cloud Storage, but I am having trouble getting the contexts and default credentials set up via the getting started docs.  The instance has access to the bucket via default credentials, so what is the proper way of setting up the code?
Also I noticed the newer API doesn't work for appengine, so I assume this library should?

Any help would be appreciated, and if needed, I'll make a new issue request.

**Comment by Capstan** [comment]
I never did use default application credentials, but my suspicion is it would look something like

```
import (
  "golang.org/x/net/context"
  "golang.org/x/oauth2"
  "google.golang.org/api/storage/v1"
)

httpClient, err := google.DefaultClient(context.Background(), storage.DevstorageFullControlScope)
svc, err := storage.New(httpClient)
```

**Comment by kamronbatman** [comment]
Thanks @Capstan!  In the past I have had problems with golang.org/x/net/context and getting an error saying "not an AppEngine context." Is there a way to get this code to work using the appengine context from the appengine go sdk that uses the custom appengine context?

**Comment by Capstan** [comment]
I think this warrants a new bug report with some code and full error information. What class is the "AppEngine context" if not a golang.org/x/net/context?

**Comment by jba** [comment]
Assigning to the developer of the next-generation code generator.

**Comment by jba** [comment]
These clients are in maintenance mode. We don't plan to address this issue.

@pongad for the next generation of clients.
